extensions/commands/recipe/cmd_create_sbom.py

```diff
@@ -0,0 +1,134 @@
+import os.path
+
+from packageurl import PackageURL
+from cyclonedx.factory.license import LicenseFactory
+from cyclonedx.model import ExternalReference, ExternalReferenceType, LicenseChoice, XsUri
+from cyclonedx.model.bom import Bom
+from cyclonedx.model.component import Component, ComponentType
+from cyclonedx.output.json import JsonV1Dot4
+
+from conan.api.output import cli_out_write, ConanOutput
+from conan.api.conan_api import ConanAPI
+from conan.cli.args import common_graph_args, validate_common_graph_args
+from conan.cli.command import conan_command
+from conans.client.graph.graph import Node
+
+lFac = LicenseFactory()
+unknown_name_created = False
+
+
+def package_type_to_component_type(pt: str) -> ComponentType:
+    if pt is "application":
+        return ComponentType.APPLICATION
+    else:
+        return ComponentType.LIBRARY
+
+
+def licenses(ids):
+    """
+    see https://cyclonedx.org/docs/1.4/json/#components_items_licenses
+    """
+    if ids is None:
+        return None
+    if not isinstance(ids, tuple):
+        ids = [ids]
+    return [LicenseChoice(license=lFac.make_from_string(i)) for i in ids]
+
+
+def name(n: Node) -> str:
+    if n.name:
+        return n.name
+    else:
+        assert globals()["unknown_name_created"] is False, "multiple nodes have no name"
+        globals()["unknown_name_created"] = True
+        return "UNKNOWN"
+
+
+def package_url(node: Node, name: str) -> PackageURL:
+    """
+    Creates a PURL following https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#conan
+    """
+    return PackageURL(
+        type="conan",
+        name=name,
+        version=node.conanfile.version,
+        qualifiers={
+            "prev": node.prev,
+            "rref": node.ref.revision if node.ref else None,
+            "user": node.conanfile.user,
+            "channel": node.conanfile.channel,
+            "repository_url": node.remote.url if node.remote else None
+        })
+
+
+def create_component(n: Node) -> Component:
+    name_ = name(n)
+    purl = package_url(n, name_)
+    result = Component(
+        type=package_type_to_component_type(n.conanfile.package_type),
+        name=name_,
+        version=n.conanfile.version,
+        licenses=licenses(n.conanfile.license),
+        bom_ref=purl.to_string(),
+        purl=purl,
+        description=n.conanfile.description
+    )
+    if n.conanfile.homepage:
+        result.external_references.add(ExternalReference(
+            type=ExternalReferenceType.WEBSITE,
+            url=XsUri(n.conanfile.homepage),
+        ))
+    return result
+
+
+def format_cyclonedx_14_json(result):
+    serialized_json = JsonV1Dot4(result).output_as_string()
+    cli_out_write(serialized_json)
+
+
+def format_text(result):
+    out = ConanOutput()
+    out.error("Format 'text' not supported")
+
+
+@conan_command(group="Recipe", formatters={
```

Comment:
The ``conan recipe:create_sbom`` reads a bit generic. (the ``recipe`` layer reads generic). I think it could make sense to have something like "sbom", "cyclonedx" as the layer name.

Not a big deal, we can move forward, keep the discussion and iterate these things, doesn't need to be agreed or changed right now.